### PR TITLE
Integrate http tunnel

### PR DESCRIPTION
**Pull request by mpetrun5**
**Short description of work done**

Added http tunnel connection with register token to load balancer

### PR Checklist

- [x] I have run linter localy
- [x] I have run unit and integration tests locally
- [x] Rebased to master branch / merged master

### Changes

-  added http tunnel to load balancer
 - changed load balancer register to return register response wiht token and tunnel_url
 - use tunnel url from register response for connecting

### Issues

Closes: #38

### CODE REVIEW

File: cmd/start.go [hunk 1]
```diff
@@ -21,7 +22,9 @@ var (
 	nodeMetricsURL string
 	id             string
 	lbBaseURL      string
+	lbTunnelURL    string
```
Comment (mpetrunic) [inline cmd/start.go:25]: eh? Ain't this suppose to be obtained from register endpoint. Whatever we do, we should have only one endpoint for load balancer in daemon

File: internal/run/run_test.go [hunk 2]
```diff
@@ -60,6 +61,7 @@ func TestStart(t *testing.T) {
 		firstGetConfigHashError     error
 		secondGetConfigHashResult   hash.Hash32
 		secondGetConfigHashError    error
+		startTunnelCallCount        int
```
Comment (mpetrunic) [inline internal/run/run_test.go:64]: ain't daemon suppose to have only one tunnel to load balancer?

**Review by mpetrunic — CHANGES_REQUESTED** [review]

File: internal/run/run_test.go (hunk 2 above)
Comment (mpetrun5) [reply internal/run/run_test.go:64]: Yep this is just to test if the function is called (0 or 1)

File: cmd/start.go (hunk 1 above)
Comment (mpetrun5) [reply cmd/start.go:25]: Well, that should be added then.
Currently this is the 5223 port that the tunnel is on

Comment (mpetrunic) [reply cmd/start.go:25]: let's add that before merging this

Comment (mpetrunic) [reply cmd/start.go:25]: cc @MakMuftic ^^

File: internal/run/run.go [hunk 3]
```diff
@@ -35,7 +35,7 @@ func Start(tunnel tunnel.Tunneler, lbClient *lb.Client, nodeClient node.Client,
 	}
 	log.Infof("Registered to load balancer %s", lbClient.BaseURL.String())
 
-	go tunnel.StartTunnel(id, registerResponse.TunnelURL, registerResponse.Token)
+	go tunnel.StartTunnel(id, registerResponse.TunnelURL, registerResponse.Token, registerResponse.Port)
```
Comment (mpetrunic) [inline internal/run/run.go:38]: change param order lel

File: internal/tunnel/tunnel.go [hunk 4]
```diff
@@ -23,17 +23,17 @@ type Tunnel struct {
...
-func (t *Tunnel) StartTunnel(nodeID string, tunnelURL string, token string) {
+func (t *Tunnel) StartTunnel(nodeID string, tunnelURL string, token string, port int) {
 	c, err := client.NewClient(&client.ClientConfig{
 		ServerAddress: tunnelURL,
 		Tunnels: map[string]*client.Tunnel{
 			"default": {
 				Protocol:   Protocol,
 				Addr:       t.NodeRPCURL.Host,
-				RemoteAddr: RemoteAddr,
+				RemoteAddr: "0.0.0.0:" + strconv.Itoa(port),
```
Comment (mpetrunic) [inline internal/tunnel/tunnel.go:36]: eh?

Comment (mpetrun5) [reply internal/tunnel/tunnel.go:36]: Well string(port) returns weird signs so this is the way to convert it string
Port is port that is returned in response that we send to assign as we can't auto assign and save to database
In vedran I will send PR with that change
So tunnelURL is url that we should connect to and port is the port we should send to be assigned on load balancer

File: internal/run/run.go (hunk 3 above)
Comment (mpetrun5) [reply internal/run/run.go:38]: Well Port is not related to TunnelURL so param order is because of that

File: cmd/start.go (hunk 1 above)
Comment (MakMuftic) [reply cmd/start.go]: This is now ready

File: internal/tunnel/tunnel.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+type Tunnel struct {
+	NodeRPCURL *url.URL
+}
+
+type Tunneler interface {
+	// StartTunnel connects to load balancer tunnel port and creates connection
+	StartTunnel(nodeID string, tunnelURL string, token string, port int)
+}
+
+func (t *Tunnel) StartTunnel(nodeID string, tunnelURL string, token string, port int) {
```
Comment (mpetrunic) [inline internal/tunnel/tunnel.go:29]: add doc comment explaining params

File: internal/tunnel/tunnel.go (hunk 4 above)
Comment (mpetrunic) [reply internal/tunnel/tunnel.go:36]: but ain't this suppose to be load balancer IP + tunnel port?

File: internal/tunnel/tunnel.go (hunk 5 above)
Comment (mpetrunic) [reply internal/tunnel/tunnel.go:29]:
```suggestion
func (t *Tunnel) StartTunnel(nodeID string, tunnelServerAddress string, authToken string, tunnelPort int) {
```

**Comment by mpetrun5** [comment]
@MakMuftic @mpetrunic 
This is still WIP 
We should first merge final tunnel PR on vedran and then I will finish this

**Review by mpetrunic — APPROVED** [review]

**Review by MakMuftic — DISMISSED** [review]
:partying_face:

**Review by mpetrunic — APPROVED** [review]

**Review by MakMuftic — APPROVED** [review]

**Pull request merged**